### PR TITLE
Implement snapshot generation

### PR DESCRIPTION
**Pull request by Jasopaum**
Fill a profile with a snapshot when differential is provided

### CODE REVIEW

File: snapshot_generation/apply_element.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    return False
+
+
+def apply_diff_element(base_element, diff_element):
+    for key_diff, val_diff in diff_element.items():
+        try:
+            func = key2func[key_diff]
+            base_element[key_diff] = func(base_element.get(key_diff), val_diff)
+        except KeyError as e:
+            print(key_diff, val_diff)
```
Comment (simonvadee) [inline snapshot_generation/apply_element.py]: do you want to keep the prints here ?

File: snapshot_generation/choice_type_element.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import copy
+import re
+
+from .apply_element import apply_diff_element_on_list
+from .composite_attribute import CompositeAttribute
+from .errors import GenerationError
+from .helper import *
+
+
+class ChoiceTypeElement(CompositeAttribute):
```
Comment (simonvadee) [inline snapshot_generation/choice_type_element.py]: I have no idea what's going on here. please add docs

File: snapshot_generation/snapshot_generator.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+from .errors import *
+from .slice import Slice
+
+
+with open("/Users/jason/Desktop/choice-elements.json", "r") as fp:
+    all_choice_elements = json.load(fp)
+
+
+class SnapshotGenerator:
+    def __init__(self):
```
Comment (simonvadee) [inline snapshot_generation/snapshot_generator.py]: I would pass the API URL to this class

File: snapshot_generation/helper.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import json
+import requests
+
+from .errors import *
+
+# TODO put that in an env variable or something like that
+API_URL = "https://pyrog.staging.arkhn.com/api"
```
Comment (simonvadee) [inline snapshot_generation/helper.py]: pass this to SnapshotGenerator class

File: snapshot_generation/helper.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def fetch_structure_definition(id_=None, url=None):
+    if not (id_ is None) ^ (url is None):
```
Comment (simonvadee) [inline snapshot_generation/helper.py:7]: didn't know this syntax, cool

Comment (simonvadee) [reply snapshot_generation/helper.py:7]: is it clearer as `if not id and not url` though ?

File: snapshot_generation/helper.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            response = requests.get(f"{API_URL}/StructureDefinition?url={url}")
+    except requests.exceptions.ConnectionError as e:
+        raise OperationOutcome(f"Could not connect to the fhir-api service: {e}")
+    if response.status_code != 200:
+        raise Exception(f"Error while fetching structure definition: {response.text}.")
+
+    return response.json()
+
+
+def fetch_extension(url):
```
Comment (simonvadee) [inline snapshot_generation/helper.py]: why do you need a function for that ? fetch_Structure_definition would do the job I think ?

File: snapshot_generation/helper.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def get_root_element(structure_definition):
+    for element in structure_definition["snapshot"]["element"]:
+        if "." not in element["id"]:
+            return element
+    raise GenerationError("root element not found in structure definition.")
+
+
+def prepend_root(root, val):
```
Comment (simonvadee) [inline snapshot_generation/helper.py:33]: what does this do ? add an example

File: snapshot_generation/snapshot_generator.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+import requests
+
+from .apply_element import *
+from .helper import *
+from .choice_type_element import ChoiceTypeElement
+from .errors import *
+from .slice import Slice
+
+
+with open("/Users/jason/Desktop/choice-elements.json", "r") as fp:
```
Comment (simonvadee) [inline snapshot_generation/snapshot_generator.py]: *cough cough*

**Review by simonvadee — COMMENTED** [review]
HEAVY FHIR, love it. However, the only thing that will make me trust this code is a proper integration test using existing StructureDefinition from HL7 for which we know both the diferential and the snapshot. We should generate the snapshot from the diff and compare the generated snapshot with the snaphsot from HL7. How does this sound ?

**Comment by Jasopaum** [comment]
> HEAVY FHIR, love it. However, the only thing that will make me trust this code is a proper integration test using existing StructureDefinition from HL7 for which we know both the diferential and the snapshot. We should generate the snapshot from the diff and compare the generated snapshot with the snaphsot from HL7. How does this sound ?

100% agree, I was going to add them, the only thing is that I don't fully agree on some snapshot/diff pairs we have...

File: snapshot_generation/apply_element.py (hunk 1 above)
Comment (Jasopaum) [reply snapshot_generation/apply_element.py]: Nope

File: snapshot_generation/choice_type_element.py (hunk 2 above)
Comment (Jasopaum) [reply snapshot_generation/choice_type_element.py]: Come on, this is crystal clear

File: snapshot_generation/helper.py (hunk 5 above)
Comment (Jasopaum) [reply snapshot_generation/helper.py:7]: I also wanted an error when both are provided (`^` is xor) but I can try to find something clearer

File: snapshot_generation/helper.py (hunk 6 above)
Comment (Jasopaum) [reply snapshot_generation/helper.py]: Totally, I forgot to clean this one, thanks 👍

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
+name: fhir.py
```
Comment (simonvadee) [inline .github/workflows/main.yml:1]: what about the publish ? we should publish the library on pypi just like fhirstore

File: snapshot_generation/choice_type_element.py (hunk 2 above)
Comment (simonvadee) [reply snapshot_generation/choice_type_element.py]: hahaha

File: .github/workflows/main.yml (hunk 9 above)
Comment (Jasopaum) [reply .github/workflows/main.yml:1]: It will make this PR quite big but I can add a publish :)

File: snapshot_generation/helper.py (hunk 5 above)
Comment (simonvadee) [reply snapshot_generation/helper.py:7]: no this is fine!

File: test/snapshot_generation/integration/test_snapshot_generation.py [hunk 10]
```diff
@@ -1,24 +1,24 @@
...
 
 # Known to have weird snapshot/differential pairs I don't agree with
 profiles_to_skip = [
-    "bp",  # FIXME Add .code on a BackboneElement?
+    "bp",  # Add .code on a BackboneElement?
     "catalog",  # One element is missing in snapshot...
     "clinicaldocument",
     "diagnosticreport-genetics",
-    "elementdefinition-de",  # FIXME List of extensions are put in reverse order
+    "elementdefinition-de",  # Extensions are extanded in snap but not in diff
```
Comment (simonvadee) [inline test/snapshot_generation/integration/test_snapshot_generation.py]: typo: extended

**Review by simonvadee — APPROVED** [review]
NICE!

**Pull request merged**